chia/util/lock.py

```diff
@@ -1,44 +1,47 @@
-import contextlib
-import os
-import time
-from typing import Iterator, Optional, TextIO, TypeVar
+from __future__ import annotations
 
-T = TypeVar("T")
+from dataclasses import dataclass
+from pathlib import Path
+from types import TracebackType
+from typing import Optional, Type
 
+from filelock import BaseFileLock, FileLock, Timeout
+from typing_extensions import final
 
-# Cribbed mostly from chia/daemon/server.py
-def create_exclusive_lock(lockfile: str) -> Optional[TextIO]:
-    """
-    Open a lockfile exclusively.
-    """
 
-    try:
-        fd = os.open(lockfile, os.O_CREAT | os.O_EXCL | os.O_RDWR)
-        f = open(fd, "w")
+class LockfileError(Exception):
+    pass
 
-        f.write("lock")
-    except IOError:
-        return None
 
-    return f
+@final
+@dataclass(frozen=True)
+class Lockfile:
```

Comment:
I think this code would be a lot simpler (and possibly more robust) if you write it as a single function with the `@contextmanager` attribute.

Then you could also use the `FileLock.acquire()` as a context manager as well.

Something like this:

```
@contextlib.contextmanager
def Lockfile(path: Path, timeout: float = -1, poll_interval: float = 0.05) -> None:
    fl = FileLock(path.with_name(path.name + ".lock"))
    try:
        using fl.aquire(timeout=timeout, poll_interval=poll_interval):
            yield
    except Timeout as e:
        raise LockfileError(e) from e
```

do you need the manual `acquire()` and `release()` calls also?

Reply:
> I think this code would be a lot simpler (and possibly more robust) if you write it as a single function with the `@contextmanager` attribute.

Thats what i had initially.

> Then you could also use the `FileLock.acquire()` as a context manager as well.

You can also just use the `FileLock` as context manager like:

```
with FileLock(path.with_name(path.name + ".lock", timeout=timeout, poll_interval=poll_interval):
    yield
```

> do you need the manual `acquire()` and `release()` calls also?

But thats ^ why i added the wrapping class because we `release` inside the context in some tests and also to make sure we have the `FileLock` creation wrapper and the exception thrown for timeouts unified and imported from one file. However, im open for suggestions to get around this with a single function `contextmanager`.

Reply:
it would be nice to be able to remove any uses that releases and re-acquires the lock while inside the context manager. Especially if it's only done in tests.